### ISSUE TITLE
TextSecure still sees encrypted SMS messages on KitKat+

### ISSUE THREAD

**Issue by johanw666** [SMSSecure/SMSSecure#6]
I have 2 phones with an identical setup: both TextSecure (2.6.4) and SecuredText (0.3.2 debug) are installed. On both phones SecurdText is set as the default sms application. Normal sms messages are received by SecuredText, however, key exchange messages and encrypted messages are received by TextSecure.

These same issues are reported with an earlier sms-only port from TextSecure: https://play.google.com/store/apps/details?id=com.securecomcode.text&hl=en ported some TextSecure 2.1 version and has the same issue. I assume the messages somehow contain information which application should handle them, this should be changed to SecuredText. I'll see if I can find more about this.

**Comment by BastienLQ** [comment]
@pR0Ps Did you test your commit in #7?

@johanw666 Can you test with the latest version of SecuredText? As TextSecure is just installed but not configured on my phone (I do not have Play Services so I cannot use TextSecure), I cannot guarantee it is fixed.

**Comment by BastienLQ** [comment]
I a fool... Tested with TextSecure v2.6.4, still not fixed.

**Comment by BastienLQ** [comment]
No, I did the test with a wrong build. It is fixed in `master`.

**Comment by johanw666** [comment]
Nice. Where can I get the new binary?

**Comment by BastienLQ** [comment]
As SecuredText is not on the Play Store or in F-Droid, you can find the latest build here: https://files.securedtext.org/org.SecuredText.SecuredText_0.3.2-debug.apk

But you really should build yourself the APK if possible.

**Comment by pR0Ps** [comment]
I'm actually still seeing this on my phone. I have TextSecure v2.6.4 and SecuredText (latest master, including the priority bump) on my phone and secured messages still pass through to TextSecure, which reports a non-existing session message.

**Comment by johanw666** [comment]
https://files.securedtext.org/org.SecuredText.SecuredText_0.3.2-debug.apk

I already have that one, that is the  version I tested it with and found the issue. And yes, I should start investing the time to build from source but my background is C++ and Delphi programming so it may take some time to setup.

**Comment by BastienLQ** [comment]
@pR0Ps Yes, TextSecure still receives secured messages, but not keys. I suspect a TextSecure bug.
@johanw666 The file has been updated. As I did not bumped version number, the file name is the same.

**Comment by johanw666** [comment]
Hmmm, odd. But since the use case is of course to combine TextSecure >= 2.7 and SecuredText, I upgraded TextSecure to 2.7.1 on both phones and installed the new debug version.

All seems to work OK now on the SecuredText end, key exchange works and I can receive and send encrypted testmessages. However, TextSecure still receives them too and complains about received encrypted sms messages and says the sender has to upgrade TextSecure. Well, it's not exactly clean but I can live with that.

Pushmessages between the same phones with TextSecure also keep working, so I now have both communication systems available in separate apps. Sad thaat it was needed this way.

**Comment by johanw666** [comment]
Hmmm. It seems that the double reception only works one-way: when I send an encrypted sms to my Android 4.4.4 device I see it both in SecuredText and TextSecure. When I send it from that phone to an Android 4.2.2 phone I see it only once, in SecuredText.

I thought that installation order might matter but I did that the same on both devices.

**Comment by pR0Ps** [comment]
The way apps access incoming SMS/MMS messages changed in KitKat.

The fact that SecuredText is seeing the message and TextSecure isn't on Jelly Bean is a good sign. It means that bumping up the priority over TextSecure worked.

I'll try messing around with it for a bit longer and see if I can prevent TextSecure from seeing the message on KitKat as well.

**Comment by johanw666** [comment]
Now that I think about it a bit more: if TextSecure and SecuredText are on the same device, I assume the use case is to use TextSecure only for pushmessages and SecuredText for both encrypted and unencrypted sms/mms. In that case TextSecure still needs to receive one kind of sms message: the registration for pushmessages. I am not sure if this is even possible to combine cleanly (instead of first installing and registering TextSecure and the installing SecuredText, in that order) without some cooperation from the TextSecure devs. But to be honest, given their attitude towards the subject, I'm not overly optimistic about that happening soon. On the other hand, they might want to help just to get the issue over with. It's worth a try.

**Comment by BastienLQ** [comment]
Does https://github.com/WhisperSystems/TextSecure/commit/472179b5500ca5e26c74b8d4f1fce1aec1872eea fix this bug?

**Comment by johanw666** [comment]
Just tested it with a self-built TextSecure that included thuis change, but an encrypted sms still is received by both applications. Should the patch in https://github.com/WhisperSystems/TextSecure/commit/472179b5500ca5e26c74b8d4f1fce1aec1872eea not also be applied to SmsListener.java ?

**Comment by pR0Ps** [comment]
WAP pushes are an MMS thing.

The check for `isDefaultSmsProvider` is already done in `SmsListener.java` as part of the `isRelevent` check so that wouldn't be it.

Some quick testing shows that `onRecieve()` is being called more than once - once where it accepts the message and calls `abortBroadcast()` and the other where it lets it through to the system (where TextSecure picks it up).

My suspicion is that this is an issue with longer messages that require more than one SMS to send.

**Comment by pR0Ps** [comment]
Closing this as it's not under our control

**Comment by pR0Ps** [comment]
On second thought, we should leave this open until the issue is resolved by the parent project.

**Comment by johanw666** [comment]
Yes,  I saw your pull request 2817 there. I built a TextSecure with this change and it works OK now, both programs work now independently without interfering. Since Moxie seems not willing to solve this (I just read his rejection) it seems I'll have to build TextSecure from source and change it myself for a while (also for the emoji pane on KitKat).

**Comment by bungabunga** [comment]
@pR0Ps is it possible to handle this problem from different perspective?

TextSecure only grabs encrypted SMS messages - it knows that a message is encrypted by the amount of blank spaces at the end of the SMS message (added by the sending client).

Is it possible for SMSSecure to label encrypted SMSes differently:
- to set a differetnt amount of blank spaces at the end of every encrypted SMS 
- to use some other method of labeling encrypted SMS messages

That way TextSecure wouldn't recognize SMSSecured messages as encrypted and wouldn't grab them in.

**Comment by pR0Ps** [comment]
The whitespace tagging is only for unencrypted messages and isn't actually used in TextSecure anymore (since the push messages already happen automatically).

The encrypted message detection is done in [WirePrefix.java](https://github.com/SMSSecure/SMSSecure/blob/master/src/org/smssecure/smssecure/protocol/WirePrefix.java) and uses (hashed) prefixes to determine the message type.

We *could* change the prefix that is used to designate a secure session. This would break compatibility with TextSecure v2.6.x (meh) and with SMSSecure before that change. The way to do it would be to look for both (involves 4000 rounds of SHA1) for the time being while sending out the new prefixes. People that use the old client would still trigger the TextSecure warning, while people that upgrade wouldn't.

After some amount of time (say a month) we would deprecate the old prefixes and show a warning if we detect an old version and later (say 5 months) deprecate it entirely and don't even look for them.

**Comment by c33s** [comment]
for me it makes sense that smssecure is compatible with textsecure < 2.7

**Comment by bungabunga** [comment]
i've solved this problem for me temporarily with Xprivacy module by not allowing TextSecure to read SMS messages and it works fine.

for this to work you have to be rooted, have Xposed framework and Xprivacy module installed.

also if you want to register with the TextSecure server you have to temporarily alow TS to get SMS messages or use a phone call registration.

**Comment by BastienLQ** [comment]
We won't break compatibility with TS <= 2.6.4. This is definitely a TS bug and workarounds in SMSSecure would imply unwanted side effects.

**Comment by Rafiot** [comment]
Maybe we could suggest TextSecure to add an option to actually disable the SMS reception 100%.

**Comment by pR0Ps** [comment]
I've hacked up something, I'll propose it in the TS mailing list and see what happens.

**Comment by a3nm** [comment]
[WhisperSystems #2817](https://github.com/WhisperSystems/TextSecure/pull/2817) was closed and locked, because the behavior was intentional to provide users time to upgrade. A new issue about the same problem has been filed with TextSecure: [WhisperSystems #4026](https://github.com/WhisperSystems/TextSecure/issues/4026).

**Comment by bungabunga** [comment]
this looks great:
https://github.com/WhisperSystems/TextSecure/compare/2951d96f0b...5aa4900b17

**Comment by johanw666** [comment]
TextSecure just released 2.28.0 (in the beta channel) with https://github.com/WhisperSystems/TextSecure/commit/92593d459b80ae0a9a275e7869a448dbf8003142 , so this issue can be closed now.

**Comment by pR0Ps** [comment]
🎉 :)
